### DISCUSSION TITLE
Apple App Store options

### DISCUSSION THREAD

**Discussion by mountainash** [web-scrobbler/web-scrobbler#3139] [Ideas]
As per #1908 I've been working on getting WebScrobbler working in Safari - I'm probably pretty close to getting it working for me (hackie code and self-compiled, unsigned install) - but I was interested to know if there's much interest (from the community/users) beyond this? It would take a bit of work to get it buildable in the existing pipeline. It would also require a [paid] Apple Developer account to distribute in the App Store. The extension could also run on iOS (& iPadOS).

### Discussion Points
- Is there much demand for this?
  - I know Google Analytics is running in the existing versions, are there number of how many users are running on Macs? [to give an indication -obviously not all Mac users use Safari]
- Would the maintainers want an Apple Dev account (or do I use my own)?
- Would the maintainers be interested in putting a price on the sale of the extension?
   - Would people be willing to pay for it?
   - How much?

The debate/answers around these questions will help me determine if it's worth investing my time into this further.

**Comment by dw808303** [comment]
I would pay.

About $2-5 feels right.

**Reply by mountainash** [reply]
I agree with that range (as a 1-off initial fee).

To cover the **$99** USD Developer annual membership fee, with a **$2** USD app (with Apple's 15% cut) would mean **37** new sales would be needed every year to cover the hard cost. At **$5** USD, it's **24** sales (~2 a month).

My initial guess is that there isn't that much interest from Safari users to cover this (maybe in the first year..., maybe).

**Comment by inverse** [comment]
First up thanks for raising this and taking it further! 

> Is there much demand for this?

Google Analytics was removed at the end of [2020](https://github.com/web-scrobbler/web-scrobbler/commit/a4d7f94875d86e9773b7fd1c7e07392127781bfd) so we really won't be able to know.

> Would the maintainers want an Apple Dev account (or do I use my own)?

Ideally we would have one that is managed by the maintainers but could probably leverage an existing one. Does Apple waver the cost for open source (free) software?

> Would the maintainers be interested in putting a price on the sale of the extension?

Personally I think the extension should be free.

I'm starting to think that migrating from https://ko-fi.com to https://opencollective.com/ would be better for the community to make it clear how things are spent and more more community driven way for budgeting for these sorts of expenses if needed.

Some other questions from my side.

Long term having it automatically publish would be amazing - like what we do have for Firefox and Chrome. I couldn't see any GitHub actions around this but are you are of any tooling for this that could be wrapped and run on GitHub Actions?

We have the same issue with Edge which often leads to that being a few versions behind due to the other maintainer's busy schedule for a similar reason.

**Reply by mountainash** [reply]
Hello @inverse - Thanks for the reply.

RE: GA removal - [the commit says that is from "Web Scrobbler v3 and newer"](https://github.com/web-scrobbler/web-scrobbler/blob/a4d7f94875d86e9773b7fd1c7e07392127781bfd/src/_locales/en/privacy.md?plain=1#L5)... but as of now the current release is v2.52.3...[^1] ? Either way, the stats would just be an indication, so if you have access to old stats and it shows that there was only 10 Mac users - then I'll just write instructions for the Wiki on how they can self-build+install. It's not worth the effort of expanding this projects support vector.

RE: Apple Dev account - from what I've seen they don't have an OSS plan, [they have a free plan](https://developer.apple.com/support/membership-fee-waiver/) for "registered" organisations, but there's lots of eligibility requirements (like having a tax number for the org in the limited set of countries on the allowed list) - ...probably easier just to pay the annual fee.

RE: Free extension - I agree that this is the best result for the user, but I don't want the project to be a financial burden. But as the financials aren't public (AFAIK), I don't know if you can cover the cost from donations. I was going to suggest moving to Open Collective if the project went down the paid App path - so I `+1` the choice of this direction.

RE: CD/CI, I'm not aware of any tools - but I'm sure there's something out there (but, like most things "Apple", there's probably a cost [as it would likely need to run on a Mac]).

[^1]: I've seen many reference to "V3", but where is it? I see [one _stale_ branch](https://github.com/web-scrobbler/web-scrobbler/tree/feature/v3) mentioning it. Everything else current is on master and it's releases are still _2.5x_. Quite confusing.

**Reply by inverse** [reply]
V3 was never released and I'm not entirely sure on the state of that branch. As I understand it was a major rewrite from pure JS to TypeScript which would be great to bring to life as it dropped a lot of the legacy baggage around RequireJS and would make the development experience a lot more modern and robust.

As for the OS split from pre-removal I just checked and Safari isn't there... 

![image](https://user-images.githubusercontent.com/713525/154835592-d09868f8-a1dd-4653-876b-8af700995cc6.png)

Having manual steps def is a good way to handle this and could be something to improve on at a later step. Thanks once again for driving this forward!

**Reply by mountainash** [reply]
Safari (the browser) would never be there....as the extension didn't run in Safari (until now).... 

But if you show the "Operating System" stats (link at the top of your screenshot) - we'd get a bit of an indication of some of the potential users (obviously many of macOS users don't use Safari, but it's a "maximum" idea to form a guess on the potential usage). Also it would be helpful to know the time period of any stats that you can share.

**Reply by inverse** [reply]
Ahh true 🙈 
![image](https://user-images.githubusercontent.com/713525/154838286-3b5fd686-6a69-4c0e-a1be-0d209286ffbc.png)


10% of the users are using MacOS it seems, but this data is really outdated based on non-updated extensions.

**Reply by mountainash** [reply]
It's higher than I thought. What time period is this? Could you get a 1 year sample from before GA was removed?

Also if you have Analytics of [the website](https://web-scrobbler.com) for actual Safari (and iOS) counts - this would also be a good indicator.

**Reply by inverse** [reply]
One year view is like 5% MacOS

![image](https://user-images.githubusercontent.com/713525/154846049-785b43f1-5a1f-4fe7-b34c-0b1c9c2c5402.png)

We don't run any analytics on the main website.

**Reply by mountainash** [reply]
Thank you very for that.... I'd say, conservatively that less than half of that would be using Safari. So for me, it wouldn't be worth the time of doing the CD and less value in doing a 1-time App Store distribution. ALSO; most of the mods made to get it working in Safari are due to Apple's different interpretation/implementation of the specs compared to the other web rendering engines & IMO Apple seems to be giving WebKit a lot of fresh love and I'm guessing/hoping much of these issues will be resolved soon - then time "could" be invested in [possibly] building a build+distribution pipeline.

**Comment by hans-juergen** [comment]
To test the waters you might also post this on the Last.fm subreddit and on the support forum in the API section of course, maybe also on the inofficial Discord server for Last.fm in the channel for scripts and tools where other developers are present, not only bot developers:

https://www.reddit.com/r/lastfm/
https://support.last.fm/c/support/api
https://discord.com/channels/260438376270921729/579673026526969876

By the way, if the Safari extension would also work on iOS, are you sure that the usual iOS restrictions for third-party apps accessing third-party sites like Last.fm would not block scrobbling e.g. YouTube in iOS Safari? That is one of the most requested features related to iOS scrobbling, as far as I can tell, as Spotify and Apple Music can already be scrobbled with iOS devices.

**Reply by dw808303** [reply]
I don’t think the restrictions on third party sites would necessarily be a problem -- there are already 3rd party music players (e.g. Marvis) that can scrobble.

**Reply by mountainash** [reply]
From what I can see; _Marvis_ is it's own .app, not a Safari Extension, so that's not the same.

@hans-juergen What's the "YouTube in Safari" issue? I'm not aware of it. YouTube works for me in Safari... is there a limitation when it's embedded within another website (I might have experienced this on Twitter)?

**Reply by hans-juergen** [reply]
@dw808303 Scrobbling in a standalone player app yes, but not with external **scrobbling** apps in iOS to **external** sites. All iOS player apps that can scrobble do this internally, also Marvis Pro and Vox, another player for macOS and iOS, also for Windows, as far as I know. But external iOS scrobbling apps like EaveScrob, QuietScrob or the official Last.fm iOS app can only scrobble local music files and Apple Music, but not other streaming sites.

@mountainash I hope I could make it clearer what I think the issue could be with YouTube scrobbling in Safari on iOS which may be different with Safari on macOS. As I don't use any of this myself, I can only collect user reports which I read on forums and try to figure out what the possible problems might be.

**Reply by mountainash** [reply]
OK there's 2 very different forms of scrobbling (well actually, _music playing detection_ being talked about here) - which you've attempted to describe. "Scrobbling" is the submission of the *detected music* to a "[scrobbler](https://github.com/web-scrobbler/web-scrobbler/tree/master/src/core/background/scrobbler)".

The thing to remember is that this project "Web Scrobbler" is only a **browser extension**/plugin (full stop). It doesn't have access to the local file system, audio sources, or other Apps - it will only capture what's playing inside the browser.

With that clarified; the Safari extension (both iOS and MacOS) would only be able to scrobbler YouTube[^1] when played within the browser (not a separate Native App installed on the device).

Could you please reword "are you sure that the usual restrictions for third-party apps accessing third-party sites like Last.fm would not block scrobbling e.g. YouTube in Safari?" as the "eg" doesn't match the question.

[^1]: Technically you could probably write a very different crafted extension that went though the _played history list_ (which may include listens made in an App outside the browser) and then submit that in bulk retrospect to the Scobbler - but that's a very different method to the way **Web Scrobbler** currently works (it detects tracks at the time of play)

**Comment by mountainash** [comment]
Ok - I have a bundled version working well enough to distribute now - how would people like it?

- 🚀 in the Apple Safari "extensions" AppStore
- ❤️ ad hoc bundled .app distribution
- 👀 diff of mods and then build/dev-install yourself (Xcode required)

Use the reaction emoji's to vote.

**Reply by inverse** [reply]
Awesome work! How much customisation was needed?

**Reply by mountainash** [reply]
The biggest issue seems to be with Safari not returning a promise for the `runtime.sendMessage` function calls - which seems to be a bug as noted in the last comment on https://developer.apple.com/forums/thread/659681 - and these calls are used in lots of places (to communicate between the page, background and popup).

There's some other small mods, but I'll raise PR's for them now as they shouldn't effect the other builds.

**Comment by bs** [comment]
@mountainash Does this live somewhere that I could pull and build to play around with now?

**Reply by mountainash** [reply]
Not at the moment. Due to lack of interest/feedback from Web Scrobber users, I didn't bother putting anything public (for fear of having to maintain it - and Safari is under rapid development again, so I hope 🤞🏼 that they'll make some changes that will allow the Firefox or Chrome builds to be used). I have a version working locally for a few months now - if you're interested I could do a diff of the changes (but it's based on the Firefox build output [not on the Web Scrobbler source])?

**Reply by bs** [reply]
That's too bad! I've personally been wanting it for some time ... I'm always listening in Safari and then realizing I need to open in another browser to track.

- I think a diff/patch of the changes would be hugely useful, even if not maintained.
- Do you have a link/ref to discussion of the work happening internally on Safari?
